packages/lit-html/src/test/lit-html_test.ts

```diff
@@ -343,4 +471,315 @@ suite('lit-html', () => {
+      render(
+        html`<a href="${'foo'}" class="bar"><div id=${'a'}></div></a>`,
+        container
+      );
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        `<a class="bar" href="foo"><div id="a"></div></a>`
+      );
+    });
+
+    test('renders to an attribute without quotes', () => {
+      render(html`<div foo=${'bar'}></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div foo="bar"></div>'
+      );
+    });
+
+    test('renders to multiple attribute expressions', () => {
+      render(
+        html`<div foo="${'Foo'}" bar="${'Bar'}" baz=${'Baz'}></div>`,
+        container
+      );
+      assert.oneOf(stripExpressionComments(container.innerHTML), [
+        '<div foo="Foo" bar="Bar" baz="Baz"></div>',
+        '<div foo="Foo" baz="Baz" bar="Bar"></div>',
+        '<div bar="Bar" foo="Foo" baz="Baz"></div>',
+      ]);
+    });
+
+    test('renders to attributes with attribute-like values', () => {
+      render(html`<div foo="bar=${'foo'}"></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div foo="bar=foo"></div>'
+      );
+    });
+
+    test('does not call a function bound to an attribute', () => {
+      const f = () => {
+        throw new Error();
+      };
+      render(html`<div foo=${f as any}></div>`, container);
+      const div = container.querySelector('div')!;
+      assert.isTrue(div.hasAttribute('foo'));
+    });
+
+    test('renders an array to an attribute', () => {
+      render(html`<div foo=${[1, 2, 3] as any}></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div foo="1,2,3"></div>'
+      );
+    });
+
+    test('renders to an attribute before a node', () => {
+      render(html`<div foo="${'bar'}">${'baz'}</div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div foo="bar">baz</div>'
+      );
+    });
+
+    test('renders to an attribute after a node', () => {
+      // prettier-ignore
+      render(html`<div>${'baz'}</div><div foo="${'bar'}"></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div>baz</div><div foo="bar"></div>'
+      );
+    });
+
+    test('renders undefined in attributes', () => {
+      render(html`<div attribute="it's ${undefined}"></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div attribute="it\'s "></div>'
+      );
+    });
+
+    test('renders undefined in attributes', () => {
+      render(html`<div attribute="${undefined}"></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div attribute=""></div>'
+      );
+    });
+
+    test('nothing sentinel removes an attribute', () => {
```

Comment:
Suggest adding tests for updates (not just initial render), specifically:
* Update to `nothing` removes attribute after having rendered one
* Update to `noChange` after having rendered something results in no DOM mutations